updating dom0 via whonix fails with qubes-vm-update

## Observation

openQA test in scenario qubesos-4.3-pull-requests-x86_64-system_tests_network_updates@64bit fails in
[TC_00_Dom0Upgrade_whonix-gateway-18](https://openqa.qubes-os.org/tests/191034/modules/TC_00_Dom0Upgrade_whonix-gateway-18/steps/2)

```
# test_000_update_qubes_vm_update
# failure: 

# timestamp 2026-08-04T03:52:35.151053
Traceback (most recent call last):
  File "/usr/lib64/python3.13/contextlib.py", line 85, in inner
    return func(*args, **kwds)
  File "/usr/lib/python3.13/site-packages/qubes/tests/integ/dom0_update.py", line 404, in test_000_update_qubes_vm_update
    self._test_000_update(self.update_func_qubes_vm_update)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/site-packages/qubes/tests/integ/dom0_update.py", line 370, in _test_000_update
    do_update_func()
    ~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/qubes/tests/integ/dom0_update.py", line 353, in update_func_qubes_vm_update
    self.fail("qubes-vm-update failed: " + f_log.read())
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: qubes-vm-update failed: 
dom0 (pending):   0%|          | 0/100 [00:00<?, ?it/s]
dom0 (updating):   0%|          | 0/100 [00:23<?, ?it/s]
dom0 (updating):   0%|          | 0.0/100 [00:23<?, ?it/s]
dom0 (updating):   0%|          | 0.0/100 [00:23<?, ?it/s]
dom0 (error):   1%|          | 1.11/100 [00:23<35:20, 21.44s/it]
dom0 (error):   1%|          | 1.11/100 [00:23<35:21, 21.45s/it]
dom0:err: Using test-inst-updatevm as UpdateVM for Dom0
dom0:err: Downloading updates. This may take a while...
dom0:out: [0;31m[0mdom0:out: Refreshing package info
dom0:out: dom0:out: Refreshing available packages.
dom0:out: dom0:err: (process:3990): librepo-WARNING **: 03:52:04.957: LRO_METALINKURL processing failed: Curl error (6): Could not resolve hostname for https://mirrors.fedoraproject.org/metalink?repo=fedora-41&arch=x86_64 [Could not resolve host: mirrors.fedoraproject.org]
dom0:out: dom0:err: Errors during downloading metadata for repository 'fedora':
dom0:out: dom0:err:   - Curl error (6): Could not resolve hostname for https://mirrors.fedoraproject.org/metalink?repo=fedora-41&arch=x86_64 [Could not resolve host: mirrors.fedoraproject.org]
dom0:out: dom0:err: Failed to download metadata for repo 'fedora': Cannot prepare internal mirrorlist: Curl error (6): Could not resolve hostname for https://mirrors.fedoraproject.org/metalink?repo=fedora-41&arch=x86_64 [Could not resolve host: mirrors.fedoraproject.org]
dom0:out: test-inst-updatevm:out: 
dom0:out: test-inst-updatevm:err: 
dom0:err: dom0 done error
dom0:out: 
dom0:err: 
```

## Test suite description



## Reproducible

Fails since (at least) Build [2026073003-devel](https://openqa.qubes-os.org/tests/190411)


## Expected result

Last good: [2026072902-devel](https://openqa.qubes-os.org/tests/190341) (or more recent)


## Further details

Always latest result in this scenario: [latest](https://openqa.qubes-os.org/tests/latest?arch=x86_64&distri=qubesos&flavor=pull-requests&machine=64bit&test=system_tests_network_updates&version=4.3)

--- a/vmupdate/agent/source/dnf/dnf5_api.py
+++ b/vmupdate/agent/source/dnf/dnf5_api.py
@@ -49,6 +49,7 @@ class DNF5(DNFCLI):
         conf = self.base.get_config()
 
         if self.type == AgentType.UPDATE_VM:
+            self.configure_whonix_maybe(conf)
             conf.config_file_path = (
                 self.UPDATE_VM_INSTALLROOT + "/etc/dnf/dnf.conf"
             )

--- a/vmupdate/agent/source/dnf/dnf_api.py
+++ b/vmupdate/agent/source/dnf/dnf_api.py
@@ -50,6 +50,7 @@ class DNF(DNFCLI):
         conf.read(filename=dnfconf)
 
         if self.type == AgentType.UPDATE_VM:
+            self.configure_whonix_maybe(conf)
             conf.best = True
             conf.plugins = False
             conf.installroot = self.UPDATE_VM_INSTALLROOT

--- a/vmupdate/agent/source/dnf/dnf_cli.py
+++ b/vmupdate/agent/source/dnf/dnf_cli.py
@@ -21,6 +21,10 @@
 
 import shutil
 from typing import List
+import os
+import random
+import string
+import subprocess
 
 from source.common.package_manager import PackageManager, AgentType
 from source.common.process_result import ProcessResult
@@ -43,6 +47,51 @@ class DNFCLI(PackageManager):
             else:
                 raise RuntimeError("Package manager not found!")
         self.package_manager: str = pck_mngr
+
+    def configure_whonix_maybe(self, conf):
+        """
+        If running inside Whonix gateway, configure DNF options as Whonix
+        normally does. Ironically, the DNFCLI class is the only one not needing
+        any extra steps, because Whonix installs a wrapper at /usr/bin/dnf, but
+        using DNF (either DNF4 or DNF5) via API needs doing it separately.
+        Place this function here, as a common place for both DNF and DNF5
+        classes.
+        """
+        # based on condition to start tor in Whonix Gateway
+        if not os.path.exists("/usr/share/anon-gw-base-files/gateway"):
+            return
+        if not os.path.exists(
+            "/run/qubes/this-is-netvm"
+        ) and not os.path.exists("/run/qubes/this-is-proxyvm"):
+            return
+
+        # now we know it's Whonix Gateway
+        # apply custom steps of dnf uwt wrapper to the conf object
+        # try to keep in sync with https://github.com/Whonix/uwt/blob/master/usr/bin/dnf-3.anondist
+
+        if (
+            shutil.which("leaprun")
+            and subprocess.call(
+                ["leaprun", "--check", "try-wait-for-tor-service-running"],
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.DEVNULL,
+            )
+            == 0
+        ):
+            subprocess.check_call(
+                ["leaprun", "try-wait-for-tor-service-running"]
+            )
+        else:
+            subprocess.check_call(
+                ["/usr/libexec/helper-scripts/try-wait-for-tor-service-running"]
+            )
+
+        random_socks_password = "".join(
+            random.choices(string.ascii_letters + string.digits, k=43)
+        )
+        conf.proxy = "socks5h://127.0.0.1:9050/"
+        conf.proxy_username = "<torS0X>0"
+        conf.proxy_password = f"dnf-wrapper_{random_socks_password}"
 
     def refresh(self, hard_fail: bool) -> ProcessResult:
         """

